lib/DxilPIXPasses/PixPassHelpers.cpp

```diff
@@ -37,15 +37,33 @@ using namespace llvm;
 using namespace hlsl;
 
 namespace PIXPassHelpers {
-bool IsAllocateRayQueryInstruction(llvm::Value const *Val) {
-  if (Val != nullptr) {
-    if (llvm::Instruction const *Inst =
-            llvm::dyn_cast<llvm::Instruction>(Val)) {
-      return hlsl::OP::IsDxilOpFuncCallInst(Inst,
-                                            hlsl::OP::OpCode::AllocateRayQuery);
+static void FindRayQueryHandlesFromUse(Value *U,
+                                       std::vector<Value *> &Handles) {
+  Handles.push_back(U);
+  auto RayQueryHandleUses = U->uses();
+  for (Use &Use : RayQueryHandleUses) {
+    iterator_range<Value::user_iterator> Users = Use->users();
+    for (User *User : Users) {
+      if (std::find(Handles.begin(), Handles.end(), User) == Handles.end())
+        FindRayQueryHandlesFromUse(User, Handles);
     }
   }
-  return false;
+}
+
+std::vector<Value *> FindRayQueryHandlesForFunction(llvm::Function *F) {
```

Comment:
Why use a `std::vector`?  Couldn't this be a `llvm::SmallPtrSet` instead?  This should be fast when small and scale well too.
Recommend using argument of type `SmallPtrSetImpl<Value*> &`, and using like so:
```cpp
llvm::SmallPtrSet<Value*, 16> RayQueryHandles;
FindRayQueryHandlesForFunction(F, RayQueryHandles);
```
That way, calling code determines the "small" size, and this function isn't locked to a particular "small" size.

Reply:
Another approach would be a helper object that handles the details for you, like a `RayQueryHandleHelper` class.  It would manage the SmallPtrSet and have an accessor for whether some value is a RayQuery handle.  That way specific details like this don't get proliferated though various parameter types and such.

Reply:
Right-o